File: leetcode/java_unittest/sortSLL_junit/sortSLLUnitTest.java
/*
 * ListNode sortSLL.mergesort(ListNode)
 *
 * listNode sortSLL.insertionsort(ListNode)
 *
 * */

import org.junit.*;
import org.junit.Assert;

class ListNode{
    int val;
    ListNode next;
    ListNode(int x){
        val = x;
        next = null;
    }
}

public class sortSLLUnitTest{
    private sortSLL game;

    @Before
    public void setUp(){
        game = new sortSLL();
    }

    @Test
    public void merge_positiveTest_1(){
        int[] arr = new int[]{10, 1, 3, 9, 7, 8, 4, 5, 2, 6};
        ListNode head = createSLL(arr);

        int[] expectedArr = new int[]{1, 2, 3, 4, 5, 6, 7, 8, 9, 10};
        ListNode expected = createSLL(expectedArr);

        ListNode result_m = game.mergesort(head);
        assertSLL(expected, result_m);
    }

    @Test
    public void merge_positiveTest_2(){
        int[] arr = new int[]{10, 1, 3, 9, 7, 8, 4, 5, 2, 6, 11};
        ListNode head = createSLL(arr);

        int[] expectedArr = new int[]{1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11};
        ListNode expected = createSLL(expectedArr);

        ListNode result_m = game.mergesort(head);
        assertSLL(expected, result_m);
    }

    @Test
    public void merge_negativeTest(){
        ListNode head = null;
        ListNode result_m = game.mergesort(head);

        Assert.assertNull(result_m);
    }

    @Test
    public void insertion_positiveTest_1(){
        int[] arr = new int[]{10, 1, 3, 9, 7, 8, 4, 5, 2, 6};
        ListNode head = createSLL(arr);

        int[] expectedArr = new int[]{1, 2, 3, 4, 5, 6, 7, 8, 9, 10};
        ListNode expected = createSLL(expectedArr);

        ListNode result_i = game.insertionsort(head);
        assertSLL(expected, result_i);
    }

    @Test
    public void insertion_positiveTest_2(){
        int[] arr = new int[]{10, 1, 3, 9, 7, 8, 4, 5, 2, 6, 11};
        ListNode head = createSLL(arr);

        int[] expectedArr = new int[]{1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11};
        ListNode expected = createSLL(expectedArr);

        ListNode result_i = game.insertionsort(head);
        assertSLL(expected, result_i);
    }

    @Test
    public void insertion_negativeTest(){
        ListNode head = null;
        ListNode result_i = game.insertionsort(head);

        Assert.assertNull(result_i);
    }

    private ListNode createSLL(int[] arr){
        ListNode head = null, h = null;
        for(int v : arr){
            if(head == null){
                head = new ListNode(v);
                h = head;
            }else{
                h.next = new ListNode(v);
                h = h.next;
            }
        }
        return head;
    }

    /*
     * to assert two SLL equals or not
     * */
    private boolean assertSLL(ListNode expected, ListNode result){
        ListNode h1 = expected, h2 = result;
        while(h1 != null || h2 != null){
            Assert.assertNotNull(h1);
            Assert.assertNotNull(h2);
            Assert.assertEquals(h1.val, h2.val);
            h1 = h1.next;
            h2 = h2.next;
        }
        return true;
    }

    @After
    public void settleDown(){
        game = null;
    }
}

File: leetcode/cpp_unittest/sortSLL_unittest/sortSLL_unittest.cpp
#include "sortSLL.cpp"
#include "gtest/gtest.h"

/*
 * sortSLL.cpp:
 * ListNode* mergesort(ListNode*)
 *
 * perliminary.cpp:
 * ListNode* createSLL(int*, int n)
 * void delSLL(ListNode*)
 *
 * */

void assertList(ListNode *expected, ListNode *result){
    ListNode *h1 = expected, *h2 = result;

    while(h1 && h2){
        EXPECT_EQ(h1->val, h2->val);
        h1 = h1->next;
        h2 = h2->next;
    }

    EXPECT_EQ(NULL, h1);
    EXPECT_EQ(NULL, h2);
}

/* ------------------ test body ---------------- */
TEST(sortSLLTest, Positive01){
    Solution s;

    int A[] = {1, 5, 7, 2, 4, 11, 3, 9, 8};
    const int n1 = sizeof(A) / sizeof(int);
    ListNode *l1 = createSLL(A, n1);

    ListNode* result = s.mergesort(l1);

    int B[] = {1, 2, 3, 4, 5, 7, 8, 9, 11};
    const int n2 = sizeof(B) / sizeof(int);
    ListNode *expected = createSLL(B, n2);

    assertList(expected, result);

    l1 = NULL;
    delSLL(expected);
    delSLL(result);
}

TEST(sortSLLTest, Positive02){
    Solution s;

    int A[] = {11, 9, 8, 7, 6, 5, 4, 3, 2, 1};
    const int n1 = sizeof(A) / sizeof(int);
    ListNode *l1 = createSLL(A, n1);

    ListNode *result = s.mergesort(l1);

    int B[] = {1, 2, 3, 4, 5, 6, 7, 8, 9, 11};
    const int n2 = sizeof(B) / sizeof(int);
    ListNode *expected = createSLL(B, n2);

    assertList(expected, result);

    l1 = NULL;
    delSLL(expected);
    delSLL(result);
}
